### PR TITLE
Add a setting to be able to always appear offline

### LINKED ISSUE #5582: Move presence enabling to an offline mode user setting
### Your use case

Remove build config for Presence enabling as its not really needed and replace it with a user setting / preference

### Have you considered any alternatives?

No

### Additional context

No

### PR DESCRIPTION
**Pull request by aringenbach**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Presence enabling has been moved to an offline mode user setting. When enabled, user will always appear offline, even when using the application.

## Motivation and context

Fixes #5582 

## Screenshots / GIFs

Screenshot_20220412_101940

## Tests

- Step 1: Have contacts on servers that show presence (e.g. local synapse)
- Step 2: Toggle setting

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s): API 32

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt [hunk 1]
```diff
@@ -71,6 +71,14 @@ class VectorSettingsPreferencesFragment @Inject constructor(
             }
         }
 
+        findPreference<VectorSwitchPreference>(VectorPreferences.SETTINGS_SHOW_PRESENCE)!!.let { pref ->
+            pref.isChecked = vectorPreferences.showPresence()
+            pref.setOnPreferenceChangeListener { _, _ ->
+                MainActivity.restartApp(requireActivity(), MainActivityArgs(clearCache = false))
```
Comment (aringenbach) [inline vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt:77]: I wonder if we can do better than restarting the MainActivity to take changes into account

**Review by aringenbach — COMMENTED** [review]
This might require input from product team, as this is adding a setting we might not want for some reason.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/handler/PresenceSyncHandler.kt [hunk 2]
```diff
@@ -31,26 +31,24 @@ import javax.inject.Inject
 internal class PresenceSyncHandler @Inject constructor(private val matrixConfiguration: MatrixConfiguration) {
 
     fun handle(realm: Realm, presenceSyncResponse: PresenceSyncResponse?) {
-        if (matrixConfiguration.presenceSyncEnabled) {
```
Comment (ariskotsomitopoulos) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/handler/PresenceSyncHandler.kt:34]: Here you removed the enable/disable logic. So now even with presence disabled the code will run.

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt (hunk 1 above)
Comment (ariskotsomitopoulos) [reply vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt:77]: Why do we need an activity restart here ? vectorPreferences.showPresence() should have the updated value after user changes the setting, it doesn't ?

File: vector/src/main/res/xml/vector_settings_preferences.xml [hunk 3]
```diff
@@ -198,6 +198,16 @@
 
     </im.vector.app.core.preference.VectorPreferenceCategory>
 
+    <im.vector.app.core.preference.VectorPreferenceCategory android:title="@string/settings_presence">
+
+        <im.vector.app.core.preference.VectorSwitchPreference
+            android:title="@string/settings_show_presence"
+            android:summary="@string/settings_show_presence_summary"
+            android:key="SETTINGS_SHOW_PRESENCE"
+            android:defaultValue="true" />
```
Comment (ariskotsomitopoulos) [inline vector/src/main/res/xml/vector_settings_preferences.xml:207]: The default value here should be the default value you defined in vector-config I guess

**Review by ariskotsomitopoulos — COMMENTED** [review]
Thanks for the PR, I added some comments

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/handler/PresenceSyncHandler.kt (hunk 2 above)
Comment (aringenbach) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/handler/PresenceSyncHandler.kt:34]: Yes this is on purpose. Since using a user setting, I wanted Realm to store Presence data regardless, so toggling is just a matter of refreshing the UI rather than having to force re-sync just to get Presence data we ignored before.

Comment (ariskotsomitopoulos) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/handler/PresenceSyncHandler.kt:34]: If its the expected behaviour its fine, thanks

**Comment by aringenbach** [comment]
> This might require input from product team, as this is adding a setting we might not want for some reason.

Done. It's ok to have this setting available, and it will be added on iOS as well as part of the presence display improvement there.

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt (hunk 1 above)
Comment (ariskotsomitopoulos) [reply vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt:77]: If the problem is for example the room list recyclerview, maybe you can try to call `requestModelBuild`  to request a model update so that epoxy will rebuild the UI for the current updated data

**Review by ariskotsomitopoulos — APPROVED** [review]
It would be good to avoid the activity restart if possible. I added a comment regarding that

File: vector-config/src/main/res/values/config-settings.xml [hunk 4]
```diff
@@ -28,6 +28,8 @@
 
     <bool name="settings_interface_bubble_visible">true</bool>
     <bool name="settings_interface_bubble_default">false</bool>
+    <bool name="settings_presence_visible">false</bool>
+    <bool name="settings_presence_offline_mode_default">false</bool>
```
Comment (aringenbach) [inline vector-config/src/main/res/values/config-settings.xml:32]: This feels worthless, I wonder if we could simply pass on this, since it kinda doubles on the default value in MatrixConfiguration 🤔

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt [hunk 5]
```diff
@@ -71,6 +76,17 @@ class VectorSettingsPreferencesFragment @Inject constructor(
             }
         }
 
+        findPreference<VectorSwitchPreference>(LightweightSettingsStorage.MATRIX_SDK_SETTINGS_PRESENCE_OFFLINE_MODE_ENABLED)!!.let { pref ->
+            pref.isChecked = lightweightSettingsStorage.getPresenceOfflineModeEnabled()
+            pref.setOnPreferenceChangeListener { _, newValue ->
+                val presenceOfflineModeEnabled = newValue as? Boolean ?: false
+                lifecycleScope.launch {
```
Comment (aringenbach) [inline vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt:83]: I wonder if there is a better place to start this but right now we don't have a viewModel or interactor on Preferences

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt [hunk 6]
```diff
@@ -71,6 +76,17 @@ class VectorSettingsPreferencesFragment @Inject constructor(
...
+                    session.setMyPresence(if (presenceOfflineModeEnabled) PresenceEnum.OFFLINE else PresenceEnum.ONLINE)
```
Comment (ariskotsomitopoulos) [inline vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt:84]: Keep in mind there is also a 3rd value `PresenceEnum.UNAVAILABLE`, that I think we currently we are not using

**Review by ariskotsomitopoulos — COMMENTED** [review]
Thanks for the update, some comments

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/MatrixConfiguration.kt [hunk 7]
```diff
@@ -61,14 +61,14 @@ data class MatrixConfiguration(
...
-         */
-        val presenceSyncEnabled: Boolean = true,
         /**
          * Thread messages default enable/disabled value
          */
         val threadMessagesEnabledDefault: Boolean = false,
+        /**
+         * Presence offline mode default enabled/disabled value
+         */
+        val presenceOfflineModeEnabledDefault: Boolean = false,
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/MatrixConfiguration.kt:71]: I think naming could maybe be a bit clearer.
What about: `userAlwaysAppearsOfflineByDefault`?

Other names could be updated using the same terminology.

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt [hunk 8]
```diff
@@ -71,6 +76,17 @@ class VectorSettingsPreferencesFragment @Inject constructor(
             }
         }
 
+        findPreference<VectorSwitchPreference>(LightweightSettingsStorage.MATRIX_SDK_SETTINGS_PRESENCE_OFFLINE_MODE_ENABLED)!!.let { pref ->
```
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt:79]: Not ideal that the app set a SharedPref that the SDK is using (IIUC). We should use the SDK API to pass settings to the SDK.

File: vector/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -1049,6 +1049,11 @@
     <string name="media_saving_period_1_month">1 month</string>
     <string name="media_saving_period_forever">Forever</string>
 
+    <!-- Presence settings -->
+    <string name="settings_presence">Presence</string>
+    <string name="settings_presence_offline_mode">Offline mode</string>
+    <string name="settings_presence_offline_mode_summary">If enabled, this client won\'t send presence online status on sync</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:1055]: It's a bit technical for the users.
I would rephrase to "If enabled, you will always appear offline to other users, even when using the application."

**Review by bmarty — CHANGES_REQUESTED** [review]
Some minor and less minor remarks

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/MatrixConfiguration.kt (hunk 7 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/MatrixConfiguration.kt:71]: (But I think this should be removed anyway)

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt (hunk 6 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/settings/VectorSettingsPreferencesFragment.kt:84]: This call to `PresenceService` should be enough to tell the SDK to use `Offline` or `Online` for every sync request when on foreground

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/presence/service/DefaultPresenceService.kt [hunk 10]
```diff
@@ -20,18 +20,22 @@ package org.matrix.android.sdk.internal.session.presence.service
...
 internal class DefaultPresenceService @Inject constructor(
         @UserId private val userId: String,
         private val setPresenceTask: SetPresenceTask,
-        private val getPresenceTask: GetPresenceTask
+        private val getPresenceTask: GetPresenceTask,
+        private val lightweightSettingsStorage: LightweightSettingsStorage
 ) : PresenceService {
 
     override suspend fun setMyPresence(presence: PresenceEnum, statusMsg: String?) {
+        lightweightSettingsStorage.setSyncPresenceStatus(SyncPresence.from(presence))
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/presence/service/DefaultPresenceService.kt:38]: Not a big deal, but storing the value could be handled by `setPresenceTask`. Not sure what is best, so let's keep it like that.

File: changelog.d/5582.change [hunk 11]
```diff
@@ -0,0 +1 @@
+Move presence enabling to an offline mode user setting
```
Comment (bmarty) [inline changelog.d/5582.change:1]: The file extension `change` is not correct and will be ignored by towncrier. Please refer to https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog

In this case, I think it's best to use the `feature` extension. For the content, since it's more user facing, I will write "Add a setting to be able to always appear offline"

**Review by bmarty — CHANGES_REQUESTED** [review]
Thanks for the update.

Just a last remark about the changelog.

Also the PR description is quite out of date. Can you update it please, for posterity?
